java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/SlidingWindowUtils.java

```diff
@@ -445,6 +448,114 @@ public static Row updateZeroValuedRow(
+     * @param rowTimeFieldName The name of the row time field.
+     * @return The result table.
+     */
+    public static Table applyGlobalWindow(
+            StreamTableEnvironment tEnv,
+            Table table,
+            SlidingWindowDescriptor windowDescriptor,
+            AggregationFieldsDescriptor aggDescriptors,
+            String rowTimeFieldName) {
+        if (windowDescriptor.stepSize != Duration.ZERO
+                || aggDescriptors.getMaxWindowSizeMs() != 0) {
+            throw new RuntimeException("Step size and window size must be 0.");
+        }
+
+        if (windowDescriptor.filterExpr != null) {
+            table = table.filter(Expressions.callSql(windowDescriptor.filterExpr));
+        }
+
+        if (windowDescriptor.limit != null) {
+            aggDescriptors =
+                    getAggregationFieldsDescriptorsWithLimit(
+                            aggDescriptors, windowDescriptor.limit);
+        }
+
+        ResolvedSchema schema = table.getResolvedSchema();
+
+        DataStream<Row> stream =
+                tEnv.toChangelogStream(
+                        table,
+                        Schema.newBuilder().fromResolvedSchema(schema).build(),
+                        ChangelogMode.insertOnly());
+
+        List<String> fieldNames = new ArrayList<>();
+        List<TypeInformation<?>> accumulatorFieldTypes = new ArrayList<>();
+        final Map<String, DataType> resultDataTypeMap = new HashMap<>();
+        for (String key : windowDescriptor.groupByKeys) {
+            fieldNames.add(key);
+            accumulatorFieldTypes.add(ExternalTypeInfo.of(getDataType(schema, key).nullable()));
+            resultDataTypeMap.put(key, getDataType(schema, key).nullable());
+        }
+        fieldNames.add(rowTimeFieldName);
+        accumulatorFieldTypes.add(ExternalTypeInfo.of(DataTypes.TIMESTAMP_LTZ(3).nullable()));
+        resultDataTypeMap.put(rowTimeFieldName, DataTypes.TIMESTAMP_LTZ(3).nullable());
+        for (AggregationFieldsDescriptor.AggregationFieldDescriptor aggDescriptor :
+                aggDescriptors.getAggFieldDescriptors()) {
+            fieldNames.add(aggDescriptor.fieldName);
+            accumulatorFieldTypes.add(
+                    aggDescriptor.aggFuncWithoutRetract.getAccumulatorTypeInformation());
+            resultDataTypeMap.put(aggDescriptor.fieldName, aggDescriptor.dataType);
+        }
+
+        final String[] keyFieldNames = windowDescriptor.groupByKeys.toArray(new String[0]);
+        final List<DataTypes.Field> resultTableFields =
+                getResultTableFields(
+                        resultDataTypeMap, aggDescriptors, rowTimeFieldName, keyFieldNames);
+        final ExternalTypeInfo<Row> resultRowTypeInfo =
+                ExternalTypeInfo.of(
+                        DataTypes.ROW(resultTableFields.toArray(new DataTypes.Field[0])));
+
+        SingleOutputStreamOperator<Row> resultStream =
+                stream.keyBy(
```

Comment:
Let's guarantee that the output parallelism for this stream is 1 in case when the group_by_keys is empty.

Could you please rebase this PR to the latest master code and add a prometheus metrics test that verifies the correctness when metric window size is 0?